### PR TITLE
Fix InventoryResult off by one error

### PR DESCRIPTION
**Pull request by insunaa**
This PR fixes a mismatch in inventory results between the legacy clients and the classic clients.

- closes https://github.com/WowLegacyCore/HermesProxy/issues/62

Edit: Tested on 1.14, too

### CODE REVIEW

File: HermesProxy/World/Server/Packets/ItemPackets.cs [hunk 1]
```diff
@@ -506,6 +506,8 @@ public class InventoryChangeFailure : ServerPacket
 
         public override void Write()
         {
+            if ((sbyte)BagResult >= 29)
+                BagResult+=1;
```
Comment (0blu) [inline HermesProxy/World/Server/Packets/ItemPackets.cs:510]:
```suggestion
                BagResult += 1;
```

**Comment by ratkosrb** [comment]
This is not how you correct enums! 

https://github.com/WowLegacyCore/HermesProxy/blob/master/HermesProxy/World/Client/PacketHandlers/ItemHandler.cs#L120
https://github.com/WowLegacyCore/HermesProxy/blob/master/HermesProxy/VersionChecker.cs#L308

**Comment by ratkosrb** [comment]
@insunaa Did you test on 1.12 or 2.4.3 core? Cause I know inventory result conversion was working correctly on vmangos. Most likely the enum is different on TBC, so go add TBC enum and use the proper conversion method.

**Comment by insunaa** [comment]
I tested on 2.4.3 core
You can check the VMaNGOS enum, it uses is (almost) identical to TBC (the exception is clients < 1.8)

**Comment by 0blu** [comment]
I still get "item not found" when trying to use the Wrapping Paper on an item in a bank.
1. Go to banker and open gosip
2. Click on Wrapping Paper
3. Open Bank
4. Click on the item in the bank

**Comment by ratkosrb** [comment]
@0blu Yes because I told you item slots need to be converted too.

https://github.com/WowLegacyCore/HermesProxy/pull/61

**Comment by insunaa** [comment]
```
19:42:11 | Error   | /mnt/980pro/git/HermesProxy/HermesProxy/World/Server/Packets/ItemPackets.cs | 128
19:42:11 | Error   | /mnt/980pro/git/HermesProxy/HermesProxy/World/Server/Packets/ItemPackets.cs | 19
19:42:11 | Error   | /mnt/980pro/git/HermesProxy/HermesProxy/World/Server/Packets/ItemPackets.cs | 31
19:42:11 | Error   | /mnt/980pro/git/HermesProxy/HermesProxy/World/Server/Packets/ItemPackets.cs | 255
19:42:11 | Error   | /mnt/980pro/git/HermesProxy/HermesProxy/World/Server/Packets/ItemPackets.cs | 47

19:43:35 | Error   | /mnt/980pro/git/HermesProxy/HermesProxy/World/Server/Packets/ItemPackets.cs | 128
19:43:35 | Error   | /mnt/980pro/git/HermesProxy/HermesProxy/World/Server/Packets/ItemPackets.cs | 19
19:43:35 | Error   | /mnt/980pro/git/HermesProxy/HermesProxy/World/Server/Packets/ItemPackets.cs | 31
19:43:35 | Error   | /mnt/980pro/git/HermesProxy/HermesProxy/World/Server/Packets/ItemPackets.cs | 77
19:43:35 | Error   | /mnt/980pro/git/HermesProxy/HermesProxy/World/Server/Packets/ItemPackets.cs | 27
```
these 2 are packets that the 2.5.2 sends when trying to use wrapping paper on an item in the bank.

I'll check what CMaNGOS expects for bank items, then

**Comment by ratkosrb** [comment]
Just look at how item slots are converted in other item packet handlers...
https://github.com/WowLegacyCore/HermesProxy/blob/master/HermesProxy/World/Server/PacketHandlers/ItemHandler.cs#L48

File: HermesProxy/World/Server/PacketHandlers/ItemHandler.cs [hunk 2]
```diff
@@ -201,10 +201,14 @@ void HandleCancelTempEnchantment(CancelTempEnchantment cancel)
         void HandleWrapItem(WrapItem item)
         {
             WorldPacket packet = new WorldPacket(Opcode.CMSG_WRAP_ITEM);
-            packet.WriteUInt8(item.GiftBag);
-            packet.WriteUInt8(item.GiftSlot);
-            packet.WriteUInt8(item.ItemBag);
-            packet.WriteUInt8(item.ItemSlot);
+            byte GiftBag = item.GiftBag != Enums.Classic.InventorySlots.Bag0 ? ModernVersion.AdjustInventorySlot(item.GiftBag) : item.GiftBag;
+            byte GiftSlot = item.GiftSlot == Enums.Classic.InventorySlots.Bag0 ? ModernVersion.AdjustInventorySlot(item.GiftSlot) : item.GiftSlot;
```
Comment (ratkosrb) [inline HermesProxy/World/Server/PacketHandlers/ItemHandler.cs:205]: You should be checking GiftBag here.

File: HermesProxy/World/Server/PacketHandlers/ItemHandler.cs [hunk 3]
```diff
@@ -201,10 +201,14 @@ void HandleCancelTempEnchantment(CancelTempEnchantment cancel)
...
+            byte ItemBag = item.ItemBag != Enums.Classic.InventorySlots.Bag0 ? ModernVersion.AdjustInventorySlot(item.ItemBag) : item.ItemBag;
+            byte ItemSlot = item.ItemSlot == Enums.Classic.InventorySlots.Bag0 ? ModernVersion.AdjustInventorySlot(item.ItemSlot) : item.ItemSlot;
```
Comment (ratkosrb) [inline HermesProxy/World/Server/PacketHandlers/ItemHandler.cs:207]: You should be checking ItemBag here.

File: HermesProxy/World/Server/PacketHandlers/ItemHandler.cs (hunk 2 above)
Comment (insunaa) [reply HermesProxy/World/Server/PacketHandlers/ItemHandler.cs:205]: What?
What does "checking GiftBag" mean?

Comment (insunaa) [reply HermesProxy/World/Server/PacketHandlers/ItemHandler.cs:205]: I have tested wrapping items in the bank and regular inventory with these changes present and everything worked exactly as expected. What do I need to check in addition to that?

Comment (ratkosrb) [reply HermesProxy/World/Server/PacketHandlers/ItemHandler.cs:205]: Go look at the examples i linked... The slot will never be equal to Enums.Classic.InventorySlots.Bag0, that's the main bag. It seems you don't understand what the code is doing. The slot only needs to be corrected if the bag is the main bag which includes equipped items, bag slots, key ring slots, and bag slots. If the bag slot is a different bag, then slots don't need to be adjusted at all.

**Comment by ratkosrb** [comment]
I fixed your code in my commit.

File: HermesProxy/World/Server/PacketHandlers/ItemHandler.cs [hunk 4]
```diff
@@ -201,10 +201,14 @@ void HandleCancelTempEnchantment(CancelTempEnchantment cancel)
...
-            packet.WriteUInt8(item.ItemBag);
-            packet.WriteUInt8(item.ItemSlot);
+            byte GiftBag = item.GiftBag != Enums.Classic.InventorySlots.Bag0 ? ModernVersion.AdjustInventorySlot(item.GiftBag) : item.GiftBag;
+            byte GiftSlot = item.GiftBag == Enums.Classic.InventorySlots.Bag0 ? ModernVersion.AdjustInventorySlot(item.GiftSlot) : item.GiftSlot;
+            byte ItemBag = item.ItemBag != Enums.Classic.InventorySlots.Bag0 ? ModernVersion.AdjustInventorySlot(item.ItemBag) : item.ItemBag;
+            byte ItemSlot = item.ItemBag == Enums.Classic.InventorySlots.Bag0 ? ModernVersion.AdjustInventorySlot(item.ItemSlot) : item.ItemSlot;
+            packet.WriteUInt8(GiftBag);
+            packet.WriteUInt8(GiftSlot);
+            packet.WriteUInt8(ItemBag);
+            packet.WriteUInt8(ItemSlot);
```
Comment (MaxtorCoder) [inline HermesProxy/World/Server/PacketHandlers/ItemHandler.cs:211]:
```suggestion
            byte giftBag = item.GiftBag != Enums.Classic.InventorySlots.Bag0 ? ModernVersion.AdjustInventorySlot(item.GiftBag) : item.GiftBag;
            byte giftSlot = item.GiftBag == Enums.Classic.InventorySlots.Bag0 ? ModernVersion.AdjustInventorySlot(item.GiftSlot) : item.GiftSlot;
            byte itemBag = item.ItemBag != Enums.Classic.InventorySlots.Bag0 ? ModernVersion.AdjustInventorySlot(item.ItemBag) : item.ItemBag;
            byte itemSlot = item.ItemBag == Enums.Classic.InventorySlots.Bag0 ? ModernVersion.AdjustInventorySlot(item.ItemSlot) : item.ItemSlot;
            packet.WriteUInt8(giftBag);
            packet.WriteUInt8(giftSlot);
            packet.WriteUInt8(itemBag);
            packet.WriteUInt8(itemSlot);
```

**Review by MaxtorCoder — CHANGES_REQUESTED** [review]

**Review by MaxtorCoder — APPROVED** [review]

**Comment by MaxtorCoder** [comment]
I fixed some code style issues 👍

File: HermesProxy/World/Server/PacketHandlers/ItemHandler.cs (hunk 2 above)
Comment (insunaa) [reply HermesProxy/World/Server/PacketHandlers/ItemHandler.cs]: I literally copy-pasted the code exactly as it was from your examples and just changed the variable names, tho?

**Pull request merged**